src/pybind/mgr/snap_schedule/module.py

```diff
@@ -37,6 +37,44 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         self._initialized = Event()
         self.client = SnapSchedClient(self)
 
+    def _subvolume_exist(self, fs: str, subvol: Union[str, None], group: Union[str, None]) -> bool:
+        rc, subvolumes, err = self.remote('volumes', 'subvolume_ls', fs, group)
+        if rc == 0:
+            for svj in json.loads(subvolumes):
+                if subvol == svj['name']:
+                    return True
+
+        return False
+
+    def _subvolume_flavor(self, fs: str, subvol: Union[str, None], group: Union[str, None]) -> str:
+        rc, subvol_info, err = self.remote('volumes', 'subvolume_info', fs, subvol, group)
+        svi_json = json.loads(subvol_info)
+        return svi_json.get('flavor', 'bad_flavor')  # 1 or 2 etc.
+
+    def _resolve_subvolume_path(self, fs: str, path: str, subvol: Union[str, None], group: Union[str, None]) -> Tuple[int, str, str]:
+        if subvol is None and group is None:
+            return 0, path, ''
+
+        rc = -1
+        subvol_path = ''
+        if self._subvolume_exist(fs, subvol, group):
+            rc, subvol_path, err = self.remote('volumes', 'subvolume_getpath', fs, subvol, group)
```

Comment:
So, instead of fetching subvolume path and then checking its flavor (version), why not fetch the subvolume snapshot path via a remote call to volumes plugin?

Reply:
the subvolume_snapshot_get_path() method returns such a path: `/volumes/_nogroup/sv1/.snap`

i.e. it includes the snapdir name as the suffix
this will add more work to sanitizing the returned name to be displayed

Reply:
since subvol snapshot base path will have the `.snap` suffix and regular dirs will not have the suffix, it will also add kludging the `create_scheduled_snapshot()` method to test if the path contains the `client_snapdir` as a suffix

if it doesn't then we'll have to append it for non-subvol dirs

Reply:
fyi - as far as my testing goes, snapshots for v1 and v2 subvolumes, both, get created immediately under the subvol dir

eg. for a v1 subvol named sv1 under the default subvolgroup, the snapshots get created under `/volumes/_nogroup/sv1/.snap` and **not** under a UUID dir

and same is the case for a v2 subvol

Reply:
its just that the `getpath` command returns different paths for v1 and v2 subvols

v1: returns `/volumes/_nogroup/sv1`
v2: returns `/volumes/_nogroup/sv2/<uuid>`

Reply:
my recommendation is to continue with existing implementation of path resolution otherwise we'll be facing too many changes that would need to be retested

Reply:
also, the current implementation is not a resource hog algorithmically

Reply:
> fyi - as far as my testing goes, snapshots for v1 and v2 subvolumes, both, get created immediately under the subvol dir

That *could* be due to auto-upgrade from v1 -> v2 iff upon first access iff the v1 subvolume didn't have any existing snapshots (under the uuid dir).

> 
> eg. for a v1 subvol named sv1 under the default subvolgroup, the snapshots get created under `/volumes/_nogroup/sv1/.snap` and **not** under a UUID dir
> 
> and same is the case for a v2 subvol

I think the v1 subvolume is getting auto-upgraded to v1 since it didn't have any existing snapshots (under the uuid) directory to begin with.

Reply:
I'll rerun the tests and check the behavior.
I think I had snapshots for the first time, but didn't have them later on.

Reply:
okay, so I think the v1 flavor of the subvolume was never tested after the introduction to v2 flavor
the reason the **Error EINVAL: error fetching subvolume metadata** exception was being thrown was because appropriate options weren't set in the `MetadataMgr` object before the `self.mark_subvolume()` method was invoked

moving the `self.mark_subvolume()` call down at an appropriate place now results in the `.meta` file to show up with the correct fields

Reply:
* `fs subvolume create_version` command is not enabled only if config key  `mgr/volumes/allow_subvolume_create_version` is enabled
* subvolume upgrade is auto-disabled if the config key  `mgr/volumes/allow_subvolume_create_version` is enabled
* qa test now randomly creates a v1 or v2 subvolumes for tests using the `create_version` command

Reply:
* currently, `allow_subvolume_upgrade` flag is not persisted across mgr restarts
This flag will need to go into the `.meta` file if it needs to be persisted